### ISSUE TITLE
pgerror: use a different type for pgcodes

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#49694]
tldr: we want to change the API signature of `pgerror.Newf` and `pgerror.Wrapf` and possibly other things that accept a pg error code as argument.

### Background

As found in #49660, there was a mistake in the code

```
  stmt := "some string"
  return pgerror.Wrapf(err, "while running %s", stmt)
```

The second argument should be the pgcode, not the error string. 

The Go type system does not catch this - the interface of `Wrapf` is `(error, string, string, ...interface{})`

It's too easy to mistakenly forget the pgcode in there.

We can't use `type Code string` or `type Code = string` because that doesn't block the error above.

### Proposal

We can do this in package `pgcode`:

```
type Code struct {
  code string
}

func MakeCode(c string) Code { return Code{c} }

func (c Code) String() string { return c.code }
```

then change `pgerror.Wrapf` to take `(error, pgcode.Code, string, ...interface{})`

All the pre-defined code can be generated to use that struct instead of a simple string (the're a generation script already, so we can automate the conversion).

suggested by @dt @tbg
